### PR TITLE
Feature/variable filtering

### PR DESCRIPTION
**Pull request by OpheliaMiralles**
## Description

Provides ability to select variables on which to compute a specific loss. Related to issue #63.

## Type of Change

-   [ ] Bug fix (non-breaking change which fixes an issue)
-   [x] New feature (non-breaking change which adds functionality)
-   [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
-   [ ] Documentation update

## Issue Number

solves #63 

## Code Compatibility

-   [x] I have performed a self-review of my code

### Code Performance and Testing

-   [ ] I have added tests that prove my fix is effective or that my feature works
-   [ ] I ran the [complete Pytest test](https://anemoi.readthedocs.io/projects/training/en/latest/dev/testing.html) suite locally, and they pass
-   [ ] I have tested the changes on a single GPU
-   [ ] I have tested the changes on multiple GPUs / multi-node setups
-   [ ] I have run the Benchmark Profiler against the old version of the code
-   [x] If the new feature introduces modifications at the config level, I have made sure to update Pydantic Schemas and default configs accordingly

### Dependencies

-   [ ] I have ensured that the code is still pip-installable after the changes and runs
-   [ ] I have tested that new dependencies themselves are pip-installable.
-   [x] I have not introduced new dependencies in the inference portion of the pipeline

### Documentation

-   [x] My code follows the style guidelines of this project
-   [x] I have updated the documentation and docstrings to reflect the changes
-   [x] I have added comments to my code, particularly in hard-to-understand areas

### CODE REVIEW

**Comment by anaprietonem** [comment]
Thanks, Ophelia, for opening this PR and [#209](https://github.com/ecmwf/anemoi-core/pull/209)!

Regarding the losses, we’re working on merging a refactor of some of these classes soon ([#52](https://github.com/ecmwf/anemoi-core/pull/52)). Once that’s merged, these PRs might need to be updated accordingly.

Happy to keep you in the loop once that happens, or if you have any questions, just let me know!

File: training/src/anemoi/training/losses/filtering.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+        squash = kwargs.get("squash", True)
+        if squash:
+            return self.loss(pred[..., self.predicted_indices], target[..., self.target_indices], **kwargs)
+        len_model_output = pred.shape[-1]
+        loss = torch.zeros(len_model_output, dtype=pred.dtype, device=pred.device, requires_grad=False)
+        loss_per_variable = self.loss(
+            pred[..., self.predicted_indices],
+            target[..., self.target_indices],
+            **kwargs,
+        )
```
Comment (HCookie) [inline training/src/anemoi/training/losses/filtering.py:106]: How does the indexing that happens here interface with the scalers?

Comment (OpheliaMiralles) [reply training/src/anemoi/training/losses/filtering.py:106]: Yes, I was waiting for a signal to adapt this code to new scalers. Is the whole refactoring done now? If yes, I'll start reworking my current open PRs.

Comment (HCookie) [reply training/src/anemoi/training/losses/filtering.py:106]: Yes, the refactoring is done. Let us know if we can be of any assistance

Comment (OpheliaMiralles) [reply training/src/anemoi/training/losses/filtering.py:106]: I think it is still useful to have target/predicted indices because variable scaling cannot deal with mapping some predicted variable (e.g. precipitation) to some data source (e.g. radar precipitation). I worked on #209 and #148 but might save time this week to refresh this PR... Thanks for your patience!

**Comment by OpheliaMiralles** [comment]
> @OpheliaMiralles could please check if the PR needs updating with main? Currently it shows 319 files changed and some pending conflicts to be resolved for some files in training, and I don't think that was the case when you opened it? Screenshot 2025-07-10 at 08 09 13 Screenshot 2025-07-10 at 08 09 44

Yes thanks should be good now

**Comment by HCookie** [comment]
Does this filtering have a use case in training, as far as I can tell it would then only constrain the targeted variable and leave the rest alone. Do you propose a use case with the `CombinedLoss`?

**Comment by OpheliaMiralles** [comment]
> Does this filtering have a use case in training, as far as I can tell it would then only constrain the targeted variable and leave the rest alone. Do you propose a use case with the `CombinedLoss`?

I can add it in the test maybe, but it is mostly useful in combination with additional target_indices in anemoi models (which are not part of this PR) so that loss can be computed against some source (NWP) or another (point observations) for the same variables

File: training/src/anemoi/training/losses/loss.py [hunk 2]
```diff
@@ -71,6 +71,12 @@ def get_loss_function(
 
     if "*" in scalers_to_include:
         scalers_to_include = [s for s in list(scalers.keys()) if f"!{s}" not in scalers_to_include]
+    if data_indices is not None:
+        kwargs.update(
+            {
+                "data_indices": data_indices,
+            },
+        )
```
Comment (HCookie) [inline training/src/anemoi/training/losses/loss.py:79]: I presume you've tested that this doesn't break the other losses?

File: training/src/anemoi/training/losses/filtering.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
+# (C) Copyright 2024 Anemoi contributors.
```
Comment (HCookie) [inline training/src/anemoi/training/losses/filtering.py]:
```suggestion
# (C) Copyright 2025- Anemoi contributors.
```

File: training/src/anemoi/training/losses/filtering.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+            self._loss_scaler_specification = ["*"]
+            self.loss = get_loss_function(DictConfig({"_target_": loss}), scalers={}, **dict(kwargs))
+        elif isinstance(loss, (DictConfig, dict)):
+            self._loss_scaler_specification = loss.pop("scalers", ["*"])
+            self.loss = get_loss_function(loss, scalers={}, **dict(kwargs))
+        elif isinstance(loss, type):
+            self._loss_scaler_specification = ["*"]
+            self.loss = loss(**kwargs)
+        else:
+            assert isinstance(loss, BaseLoss)
```
Comment (HCookie) [inline training/src/anemoi/training/losses/filtering.py]: Another if and a proper exception would be better here

File: training/src/anemoi/training/losses/loss.py (hunk 2 above)
Comment (OpheliaMiralles) [reply training/src/anemoi/training/losses/loss.py:79]: I launched the full anemoi training test set so it should not

File: training/src/anemoi/training/losses/scalers/variable_masking.py [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+        variables: list[str],
+        data_indices: IndexCollection,
+        metadata_extractor: ExtractVariableGroupAndLevel,
+        inverse: bool = False,
+        norm: str | None = None,
+        **kwargs,
+    ) -> None:
+        weights = dict.fromkeys(variables, 0.0) if not inverse else dict.fromkeys(variables, 1.0)
+        weights["default"] = 1.0 if not inverse else 0.0
+        super().__init__(data_indices, DictConfig(weights), metadata_extractor, norm, **kwargs)
```
Comment (HCookie) [inline training/src/anemoi/training/losses/scalers/variable_masking.py]:
```suggestion
class VariableMaskingLossScaler(GeneralVariableLossScaler):
    """Class for masking variables out in the loss.
    
    Use `invert=True` to set this to only mask the specified
    variables in.
    """

    def __init__(
        self,
        variables: list[str],
        data_indices: IndexCollection,
        metadata_extractor: ExtractVariableGroupAndLevel,
        invert: bool = False,
        norm: str | None = None,
        **kwargs,
    ) -> None:
        weights = dict.fromkeys(variables, 0.0) if not invert else dict.fromkeys(variables, 1.0)
        weights["default"] = 1.0 if not invert else 0.0
        super().__init__(data_indices, DictConfig(weights), metadata_extractor, norm, **kwargs)
```

File: training/tests/unit/train/test_loss_scaling.py [hunk 6]
```diff
@@ -260,3 +296,41 @@ def test_metric_range(fake_data: tuple[DictConfig, IndexCollection]) -> None:
...
+        statistics_tendencies=statistics_tendencies,
+        metadata_extractor=metadata_extractor,
+        output_mask=NoOutputMask(),
+    )
+    vars_to_mask = ["z", "other", "q"]
+    indices_to_mask = [data_indices.model.output.name_to_index[v] for v in vars_to_mask]
+    assert scalers["variable_masking"][0][0] == len(vars_to_mask)
+    assert not scalers["variable_masking"][1][indices_to_mask].any(), "Expected scalers for masked variables to be zero"
+
+    config.training.scalers.builders["variable_masking"].update(inverse=True)
```
Comment (HCookie) [inline training/tests/unit/train/test_loss_scaling.py]:
```suggestion
    config.training.scalers.builders["variable_masking"].update(invert=True)
```

File: training/src/anemoi/training/losses/scalers/variable_masking.py [hunk 7]
```diff
@@ -0,0 +1,42 @@
+# (C) Copyright 2024 Anemoi contributors.
```
Comment (HCookie) [inline training/src/anemoi/training/losses/scalers/variable_masking.py]:
```suggestion
# (C) Copyright 2025- Anemoi contributors.
```

**Review by HCookie — APPROVED** [review]
